File: desktop/src/services/pollinate.ts
/**
 * Pollinate Content-Asset Gallery API service.
 *
 * ASSET-CENTRIC (sibling of pipelines.ts, but the first-class object is a produced
 * media asset, not a run). Fetches the newest-first content cards + overall rollup
 * from GET /api/pollinate/assets, and one topic's detail from GET /api/pollinate/{run}.
 *
 * Thumbnails: the frontend points <img> at the EXISTING GET /api/workspace/file/raw
 * (workspace-sandboxed FileResponse) — see assetThumbUrl(). No media bytes flow
 * through this service.
 *
 * Exports:
 * - pollinateService — { fetchAssets(), fetchTopicDetail(run) }
 * - assetThumbUrl(path) — absolute /api/workspace/file/raw URL for an <img src>
 * - camelCase types: PollinateAsset, PollinateContentCard, PollinateOverall, ...
 */

import api from './api';
import { getApiBaseUrl } from './tauri';

export interface PollinateAsset {
  platform: string;
  format: string;
  filePath: string;
  fileName: string;
  isImage: boolean;
  publishStatus: 'ready' | 'ready-to-publish' | 'published';
  /** Stable logical id (sha1 of platform/format/fileName) — the key for POST /publish. */
  assetId: string;
  /** Public URL where the asset was posted (set once marked published). */
  postedUrl: string | null;
}

export interface PollinateContentCard {
  run: string;
  topic: string;
  domain: string | null;
  status: string;
  createdAt: string | null;
  hasRunJson: boolean;
  assetCount: number;
  platforms: string[];
  formats: string[];
  publishedCount: number;
  readyCount: number;
  assets: PollinateAsset[];
}

export interface PollinateOverall {
  cardCount: number;
  assetCount: number;
  platformDist: Record<string, number>;
  formatDist: Record<string, number>;
  domainDist: Record<string, number>;
  published: number;
  ready: number;
  inProgress: number;
  /** Known-channel universe (server SSOT) — lets Insights grey out fully-neglected channels. */
  knownChannels: string[];
}

export interface PollinateAssetsResponse {
  overall: PollinateOverall;
  cards: PollinateContentCard[];
}

export interface PollinateTopicDetail {
  run: string;
  topic: string;
  domain: string | null;
  status: string;
  createdAt: string | null;
  contentPackage: string | null;
  assets: PollinateAsset[];
}

function assetToCamel(a: Record<string, unknown>): PollinateAsset {
  return {
    platform: (a.platform as string) ?? '',
    format: (a.format as string) ?? 'other',
    filePath: (a.file_path as string) ?? '',
    fileName: (a.file_name as string) ?? '',
    isImage: Boolean(a.is_image),
    publishStatus: (a.publish_status as PollinateAsset['publishStatus']) ?? 'ready',
    assetId: (a.asset_id as string) ?? '',
    postedUrl: (a.posted_url as string | null) ?? null,
  };
}

function cardToCamel(c: Record<string, unknown>): PollinateContentCard {
  return {
    run: c.run as string,
    topic: (c.topic as string) ?? (c.run as string),
    domain: (c.domain as string | null) ?? null,
    status: (c.status as string) ?? 'unknown',
    createdAt: (c.created_at as string | null) ?? null,
    hasRunJson: Boolean(c.has_run_json),
    assetCount: (c.asset_count as number) ?? 0,
    platforms: (c.platforms as string[]) ?? [],
    formats: (c.formats as string[]) ?? [],
    publishedCount: (c.published_count as number) ?? 0,
    readyCount: (c.ready_count as number) ?? 0,
    assets: Array.isArray(c.assets) ? (c.assets as Record<string, unknown>[]).map(assetToCamel) : [],
  };
}

function overallToCamel(o: Record<string, unknown>): PollinateOverall {
  return {
    cardCount: (o.card_count as number) ?? 0,
    assetCount: (o.asset_count as number) ?? 0,
    platformDist: (o.platform_dist as Record<string, number>) ?? {},
    formatDist: (o.format_dist as Record<string, number>) ?? {},
    domainDist: (o.domain_dist as Record<string, number>) ?? {},
    published: (o.published as number) ?? 0,
    ready: (o.ready as number) ?? 0,
    inProgress: (o.in_progress as number) ?? 0,
    knownChannels: (o.known_channels as string[]) ?? [],
  };
}

/** Absolute /api/workspace/file/raw URL for a workspace-relative asset path. Used
 *  as an <img src> — mirrors MarkdownRenderer's image-URL construction. */
export function assetThumbUrl(relativePath: string): string {
  return `${getApiBaseUrl()}/api/workspace/file/raw?path=${encodeURIComponent(relativePath)}`;
}

export const pollinateService = {
  async fetchAssets(): Promise<PollinateAssetsResponse> {
    const { data } = await api.get('/pollinate/assets');
    const d = data as Record<string, unknown>;
    return {
      overall: overallToCamel((d.overall as Record<string, unknown>) ?? {}),
      cards: Array.isArray(d.cards) ? (d.cards as Record<string, unknown>[]).map(cardToCamel) : [],
    };
  },

  /** Lazily read ONE text asset's body (caption/narrative) on drawer-open — via the
   *  existing workspace file endpoint (returns {content, encoding}). Kept OUT of
   *  fetchAssets so the gallery load stays fast (the 87ms baseline is untouched). */
  async fetchAssetBody(path: string): Promise<string | null> {
    try {
      const { data } = await api.get('/workspace/file', { params: { path } });
      const d = data as Record<string, unknown>;
      if (d.encoding === 'utf-8' && typeof d.content === 'string') return d.content;
      return null; // binary/other → not renderable as caption text
    } catch {
      return null;
    }
  },

  async fetchTopicDetail(run: string): Promise<PollinateTopicDetail | null> {
    try {
      const { data } = await api.get(`/pollinate/${encodeURIComponent(run)}`);
      const d = data as Record<string, unknown>;
      return {
        run: d.run as string,
        topic: (d.topic as string) ?? (d.run as string),
        domain: (d.domain as string | null) ?? null,
        status: (d.status as string) ?? 'unknown',
        createdAt: (d.created_at as string | null) ?? null,
        contentPackage: (d.content_package as string | null) ?? null,
        assets: Array.isArray(d.assets) ? (d.assets as Record<string, unknown>[]).map(assetToCamel) : [],
      };
    } catch {
      return null;
    }
  },

  /** Mark ONE asset published/unpublished (P1 write path). Persists to the run's
   *  publish-state.json sidecar. Returns the new publish_status, or null on failure
   *  (caller keeps the old UI state). */
  async markPublished(
    run: string, assetId: string, published: boolean, postedUrl?: string | null,
  ): Promise<{ publishStatus: PollinateAsset['publishStatus']; postedUrl: string | null } | null> {
    try {
      const { data } = await api.post(`/pollinate/${encodeURIComponent(run)}/publish`, {
        asset_id: assetId,
        published,
        posted_url: postedUrl || null,
      });
      const d = data as Record<string, unknown>;
      return {
        publishStatus: (d.publish_status as PollinateAsset['publishStatus']) ?? 'ready',
        postedUrl: (d.posted_url as string | null) ?? null,
      };
    } catch {
      return null;
    }
  },
};

File: backend/schemas/pollinate.py
"""Pydantic response models for the Pollinate Content-Asset Gallery.

Sibling of `schemas/pipeline_run.py`, but ASSET-CENTRIC: the first-class object is a
produced media asset (a file under a run's `deliver/` tree), not the run. A pollinate
run's `run.json` is optional metadata — the `deliver/` walk is the source of truth for
what was actually produced (verified: 5/14 run dirs have run.json, 10/14 have deliver/,
and `run.json.formats/platforms` are often None on running runs while the deliver/ tree
holds the real files). See the design doc:
Knowledge/Designs/2026-08-02-pollinate-content-asset-gallery-navcard-design.md
"""

from __future__ import annotations

from typing import Optional
from urllib.parse import urlsplit

from pydantic import BaseModel, Field, field_validator


class PollinateAsset(BaseModel):
    """One produced deliverable file under `deliver/[{platform}/]{file}`."""

    platform: str = Field(description="xiaohongshu / bilibili / github / gongzhonghao / '' (bare)")
    format: str = Field(description="poster / narrative / video / caption / readme / qr / other")
    file_path: str = Field(description="workspace-relative path (for /api/workspace/file/raw)")
    file_name: str
    is_image: bool = Field(description="True → renderable via <img src=/api/workspace/file/raw>")
    publish_status: str = Field(
        default="ready",
        description="ready (default/unknown) | ready-to-publish (kit present) | published",
    )
    asset_id: str = Field(
        default="",
        description="Stable LOGICAL id = sha1(platform/format/file_name). The key the "
        "publish-state sidecar uses + the frontend passes to POST /publish. Stable across "
        "the deliver↔tracks root-flip (a physical-path key would orphan the mark).",
    )
    posted_url: Optional[str] = Field(
        default=None, description="The public URL where this asset was posted (set when published)."
    )


class PublishRequest(BaseModel):
    """`POST /api/pollinate/{run_name}/publish` body — mark one asset published/unpublished."""

    asset_id: str = Field(description="The asset's stable logical id (40-hex sha1).")
    published: bool = Field(description="True = mark published; False = revert to ready.")
    posted_url: Optional[str] = Field(
        default=None, max_length=2048, description="Optional public URL of the post (http/https only)."
    )

    @field_validator("posted_url")
    @classmethod
    def _validate_posted_url(cls, v: Optional[str]) -> Optional[str]:
        """Reject a non-http(s) URL. posted_url is stored verbatim and later rendered as an
        <a href> in the overlay — an un-scheme-checked value permits a stored javascript:/data:
        XSS (there is no CSP). Allow only http/https; empty → None. (Security-gate, 422 on fail.)"""
        if v is None:
            return None
        v = v.strip()
        if not v:
            return None
        parts = urlsplit(v)
        # Require BOTH an http(s) scheme AND a network location (netloc) — i.e. the full
        # `http(s)://host` form. Checking scheme alone would accept `https:x` / `http:foo`
        # (no `//`), which the frontend render guard (/^https?:\/\//) rejects → a stored value
        # the UI silently won't render. Requiring netloc makes backend + frontend agree.
        if parts.scheme.lower() not in ("http", "https") or not parts.netloc:
            raise ValueError("posted_url must be a full http:// or https:// URL")
        return v


class PublishResponse(BaseModel):
    """`POST /api/pollinate/{run_name}/publish` result — the asset's new publish state."""

    asset_id: str
    publish_status: str = Field(description="'published' or 'ready' after the write.")
    posted_url: Optional[str] = None


class PollinateContentCard(BaseModel):
    """One content topic (one pollinate run) + all the assets it produced."""

    run: str = Field(description="run dir name, e.g. 2026-05-03-memory-is-the-moat")
    topic: str = Field(description="topic || message || run dir name (fallback chain)")
    domain: Optional[str] = None
    status: str = Field(default="unknown", description="run.json status, or 'unknown' if no run.json")
    created_at: Optional[str] = None
    has_run_json: bool = Field(default=False, description="False → card synthesized from deliver/ only")
    asset_count: int = 0
    platforms: list[str] = Field(default_factory=list)
    formats: list[str] = Field(default_factory=list)
    published_count: int = 0
    ready_count: int = 0
    assets: list[PollinateAsset] = Field(default_factory=list)


class PollinateOverall(BaseModel):
    """Top-of-gallery rollup across all content cards."""

    card_count: int = 0
    asset_count: int = 0
    platform_dist: dict[str, int] = Field(default_factory=dict)
    format_dist: dict[str, int] = Field(default_factory=dict)
    domain_dist: dict[str, int] = Field(default_factory=dict)
    published: int = 0
    ready: int = 0
    in_progress: int = Field(default=0, description="cards whose run status is running/review")
    known_channels: list[str] = Field(
        default_factory=list,
        description="The known-channel UNIVERSE (from the server's _KNOWN_PLATFORMS SSOT). "
        "Lets the Insights by-channel view surface a FULLY-neglected channel (0 assets "
        "anywhere) that platform_dist alone can never reveal — without the frontend "
        "hardcoding a drift-prone duplicate list.",
    )


class PollinateAssetsResponse(BaseModel):
    """`GET /api/pollinate/assets` — newest-first content cards + overall rollup.

    `overall` is computed over the WHOLE corpus (all counts honest). `cards` is the
    newest-N PAYLOAD slice (the cap bounds only the returned list size, not the
    rollup) — the frontend renders newest-first, so older cards beyond the cap are
    reflected in `overall` but not shipped as individual cards.
    """

    overall: PollinateOverall
    cards: list[PollinateContentCard] = Field(default_factory=list)


class PollinateTopicDetail(BaseModel):
    """`GET /api/pollinate/{run_name}` — one content topic's full detail."""

    run: str
    topic: str
    domain: Optional[str] = None
    status: str = "unknown"
    created_at: Optional[str] = None
    content_package: Optional[str] = Field(default=None, description="content_package.md body if present")
    assets: list[PollinateAsset] = Field(default_factory=list)
